Fix indexing_dtype_strength_reduction

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91601

Many of the previous inductive cases were wrong (e.g. `abs`, `sq`, `div` and `truediv`). 
We rewrite it using the mathematical terms that allow to prove the relevant upper 
and lower bounds.

Note that the inductive step can be seen as a not-too-difficult optimisation problem
with constraints, hence the naming of the functions.

For many of the other functions, we also simplify the formulas, which will be useful
when this code is generalised to work with symbolic shapes.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

lezcano committed:
Fix indexing_dtype_strength_reduction

The previous inductive cases were completely wrong. We rewrite it using
the mathematical terms that allow us to prove the relevant upper and
lower bounds.

[ghstack-poisoned]

lezcano committed:
Update on "Fix indexing_dtype_strength_reduction"


The previous inductive cases were completely wrong. We rewrite it using
the mathematical terms that allow us to prove the relevant upper and
lower bounds.

Note that the inductive step can be seen as a not-too-difficult optimisation problem
with constraints, hence the naming of the functions.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Fix indexing_dtype_strength_reduction"


Many of the previous inductive cases were wrong (e.g. `abs`, `sq`, `div` and `truediv`). 
We rewrite it using the mathematical terms that allow to prove the relevant upper 
and lower bounds.

Note that the inductive step can be seen as a not-too-difficult optimisation problem
with constraints, hence the naming of the functions.

For many of the other functions, we also simplify the formulas, which will be useful
when this code is generalised to work with symbolic shapes.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented on torch/_inductor/optimize_indexing.py:
```diff
@@ -136,77 +156,94 @@ def constant(value, dtype):
+        out = ValueRangeAnalysis.truediv(a, b)
+        floor = sympy.functions.elementary.integers.floor
+        ceil = sympy.functions.elementary.integers.ceiling
+        return ValueRanges(ceil(out.lower), floor(out.upper))
```
`//` is `floor(a/b)` so I should use `floor` on both bounds. I'll fix that.

lezcano committed:
Update on "Fix indexing_dtype_strength_reduction"


Many of the previous inductive cases were wrong (e.g. `abs`, `sq`, `div` and `truediv`). 
We rewrite it using the mathematical terms that allow to prove the relevant upper 
and lower bounds.

Note that the inductive step can be seen as a not-too-difficult optimisation problem
with constraints, hence the naming of the functions.

For many of the other functions, we also simplify the formulas, which will be useful
when this code is generalised to work with symbolic shapes.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel reviewed (approved):
Seems reasonable to me.

@eellison can you take a look too?

eellison commented on torch/_inductor/optimize_indexing.py:
```diff
@@ -18,8 +18,12 @@
+    lower: Union[sympy.Expr, sympy.Number, int, float, bool]
+    upper: Union[sympy.Expr, sympy.Number, int, float, bool]
+
+    def __contains__(self, x):
```
This might be a bit of a confusing api, because it's not necessarily going to return True or False as is the typical convention, it might return a sympy expression. maybe assert that the output is not an expr?

eellison commented on torch/_inductor/optimize_indexing.py:
```diff
@@ -136,77 +156,105 @@ def constant(value, dtype):
 
     @staticmethod
     def reciprocal(x):
-        return ValueRanges.checked_unary_map(x, lambda y: 1 / y)
+        x = ValueRanges.wrap(x)
+        if 0 in x:
+            return ValueRanges(-math.inf, math.inf)
+        else:
+            return ValueRanges.decreasing_map(x, lambda y: 1 / y)
```
would be simpler as `V.get_ops_handler().div(1, y)` i think

lezcano replied:
What do you mean exactly?

eellison replied:
reciprocal is `div(1, x)`, we could reuse div instead of duplicating the function

lezcano replied:
ah, the thing is that this specialisation gives simpler expressions for the case of symbolic shapes, that's why I kept it

eellison commented on torch/_inductor/optimize_indexing.py:
```diff
@@ -136,77 +156,105 @@ def constant(value, dtype):
     def div(a, b):
-        return ValueRanges.binary_map_products(a, b, operator.truediv)
+        # We think of this as floor(a / b)
+        out = ValueRangeAnalysis.truediv(a, b)
+        floor = sympy.functions.elementary.integers.floor
+        return ValueRanges(floor(out.lower), floor(out.upper))
```
I dont think this is what we want if the inputs are floating point. could you also check that sympy derivatives are able to handle floor ?

lezcano replied:
I restructured a bit this function and `floor` and `ceil` for their implementation to be a bit simpler and correct.

eellison commented on torch/_inductor/optimize_indexing.py:
```diff
@@ -136,77 +156,105 @@ def constant(value, dtype):
     def pow(a, b):
-        return ValueRanges.binary_map_products(a, b, operator.pow)
+        def is_integer(val):
+            return isinstance(val, int) or (
```
we might want to update the `to_dtype` to accurately reflect dtype conversions if we are depending on it.

lezcano replied:
Where are we depending on `to_dtype`?

eellison replied:
we're not depending on `to_dtype`, but we are depending on whether the a value is an int or float.

eellison commented on torch/_inductor/optimize_indexing.py:
```diff
@@ -136,77 +156,105 @@ def constant(value, dtype):
-        return ValueRanges.binary_map_products(a, b, operator.pow)
+        def is_integer(val):
+            return isinstance(val, int) or (
+                hasattr(val, "is_integer") and val.is_integer
```
should we also check if floating point values have no decimal ?

eellison reviewed (commented):
nice ! left a few comments. i think mostly good to land although we don't want the floor on `div` for floating points

lezcano committed:
Update on "Fix indexing_dtype_strength_reduction"


Many of the previous inductive cases were wrong (e.g. `abs`, `sq`, `div` and `truediv`). 
We rewrite it using the mathematical terms that allow to prove the relevant upper 
and lower bounds.

Note that the inductive step can be seen as a not-too-difficult optimisation problem
with constraints, hence the naming of the functions.

For many of the other functions, we also simplify the formulas, which will be useful
when this code is generalised to work with symbolic shapes.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Fix indexing_dtype_strength_reduction"


Many of the previous inductive cases were wrong (e.g. `abs`, `sq`, `div` and `truediv`). 
We rewrite it using the mathematical terms that allow to prove the relevant upper 
and lower bounds.

Note that the inductive step can be seen as a not-too-difficult optimisation problem
with constraints, hence the naming of the functions.

For many of the other functions, we also simplify the formulas, which will be useful
when this code is generalised to work with symbolic shapes.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison reviewed (approved):
lgtm - additional tests would b good

lezcano commented:
@pytorchbot merge -f "test errors are unrelated"

ezyang commented:
Did any tests end up ever getting written

lezcano commented:
I didn't. I don't know if @eellison did.
